### ISSUE TITLE
Better way to handle poisoned message with message ordering?

### ISSUE THREAD

**Issue by kendallb** [rebus-org/Rebus#1067]
We ran into an issue today where we had some new code that was deploying across our cluster that ended up posting some badly formatted messages onto the message bus. Well not badly formatted, we added some new fields to one of the messages and when old code that was still running on nodes that had not yet been upgraded posted the message without the new fields, they were then null. So when the message queue attempted to process them they crashed out and then ended up in the poisoned message error queue. 

But itself that is not a problem, except once all the code was correctly deployed, we had failed messages in the error queue that could not be played (not until we added code to ignore the missing data). But since messages posted on the bus from nodes that had been upgraded AFTER the bad ones got punted to the error queue happened, now we have a potential for data loss as the original messages still need to get played, but messages posted after that were not blocked and could well have messed things up.

So thinking about this I think the correct solution is to somehow block messages in the message queue from being processed if they have the same ordering key as one in the poison queue? But what is the best way to do that? Since the error queue and message queue are separate and they do not really know about each other, it's not really feasible to add something to do that. So I am thinking maybe there are two clean solutions here:

1. Add the ability to automatically punt messages to the error queue when they are posted to the bus if anything exists in the poison queue already for that ordering key? To do that somehow the message queue would need to know about the error handler and be able to find out of a message exists there? Or maybe it can be done in the generic Rebus code when it goes to publish the message, to first check if something with that ordering key already exists?

2. Have some way to keep the poisoned messages in the normal message queue and simply flag them as poisoned so they do not get played? Then the code that would normally handle the ordering keys would automatically get blocked on the first poisoned message until whatever caused it it get stuck is fixed and it gets replayed?

Thoughts?

One caveat is that ordering key support is not generic, it is part of the specific Rebus transport modules, in my case the MySQL one. Its part of the header for the message so not sure how this would work in such as way that it could be done in a generic fashion so it could be used for any message handler?

**Comment by mookid8000** [comment]
I have thoughts about this 🙂 

The thing is this: The general advice is to _always_ be sure that queue-based systems are compatible at least one version back, because you will almost always experience having to process messages sent to queues by the previous version (and sometimes maybe even one or more versions before that, when you factor in rapid, automated deployments, dead-letter queues, and possibly autonomous teams and/or weekends/other differences in when teams are working 😉 )

So I would be very reluctant to try to offer any "help" with solving this issue by adding features to Rebus and/or its transports, because I think you should work hard to avoid having this problem in the first place.

In your case, I suggest you deploy a temporary little Rebus endpoint that consumes the messages from the dead-letter queue, corrects them, and then forwards them to their source queues (possibly just cycling through them, putting them back at the end of the queue with `await bus.Advanced.TransportMessage.Forward("error");` if they're not supposed to be corrected).

How do you think that sounds?

**Comment by kendallb** [comment]
Yes I agree is best to avoid problems like this but you can’t always achieve that. In this case it was due to a message version and we missed that going to production. We have already done a deep dive on the cause of the problem and made changes to that part of the code to better avoid that in the future.

But the real problem is you cannot always avoid the crashes. What if the scenario was not a message version problem, but a system acting up? The entire point of us using rebus is to make our systems more resilient to failure, so for instance if our database or cache tier crapped out and messages failed to be processed long enough that they ended up in the dead letter queue, once the database is restored messages can process again but the dead letters will not as right now it’s a manual process for us to re-release them all. But until someone does we don’t want messages with the same ordering keys to get processed at all. Even if we had some kind of auto retry strategy for the dead letter queue we cannot allow messages to process until those dead ones are fixed and cleared for flight.

I know message ordering is not part of the base rebus and I implemented support for it in the MySQL transport but it is supported by a lot of other transports. So if the dead letter queue is separate there, then you would have a similar problem unless that transport would guarantee ordering across the normal message queue and the dead letter one? Do any transports currently do that?

I am leaning towards adding support in the MySQL transport to have the error messages live in the normal queue with a flag to indicate it’s a dead one. So then the ordering works and messages back up until we fix things and release them.

Or maybe a better solution is to simply not have a dead letter queue and at all and have the messages stay in the main message queue and we just keep trying them? I already have a jittered back off strategy to retry them but it does eventually give up and punt to the dead letter queue. So maybe the right solution is to keep that setup but top out at a decent retry rate (slow enough not to swamp things and not so slow it would take too long to recover), and then change our message visualization tool to see ones that are stuck in the normal transport (and probably log to our error logs when messages are failing so we know it’s happening).

Tempted to go with the latter…

Speaking of message versioning, how do you recommend someone implement something like that? I do like that idea but need to come up with a way to handle it.

**Comment by mookid8000** [comment]
> (..) Or maybe a better solution is to simply not have a dead letter queue and at all and have the messages stay in the main message queue and we just keep trying them? 

You can effectively achieve this by running with parallelism = 1 and enabling automatic lease renewal

```csharp
services.AddRebus(
	configure => configure
		.Transport(t => t.UseSqlServerInLeaseMode(..., automaticallyRenewLeases: true))
		.Options(o => o.SetMaxParallelism(1))
);
```

(and not running any additional instances, obviously 😅 ), and then let your message handler retry indefinitely (via [Polly](https://www.nuget.org/packages/Polly)):

```csharp
public class MyMessageHandler : IHandleMessages<MyMessage>
{
	static readonly AsyncPolicy RetryPolicy = Policy
		.Handle<Exception>()
		.WaitAndRetryForeverAsync(_ => TimeSpan.FromSeconds(1));

	readonly IMessageContext _messageContext;

	public MyMessageHandler(IMessageContext messageContext) => _messageContext = messageContext;

	public async Task Handle(MyMessage message)
	{
		var cancellationToken = _messageContext.GetCancellationToken();

		await RetryPolicy.ExecuteAsync(token => InnerHandle(message, token), cancellationToken);
	}

	async Task InnerHandle(MyMessage message, CancellationToken cancellationToken)
	{
		// (...)
	}
}
```

But, as you can probably imagine, this will severely cripple your ability to scale, as there will be NO parallel processing of anything, and nothing can continue if there's an exception.

**Comment by kendallb** [comment]
> You can effectively achieve this by running with parallelism = 1 and enabling automatic lease renewal

Yeah but as you said it kills parallelism and we need multiple nodes for fault tolerance and also code deploys :) Right now when we deploy need code we shut down one Rebus node at a time so there is always something there to process messages.

Either way fixing poisoned message handling is not the solution so we are doing it by making the message processing itself better :)
